edgelet/iotedged/src/lib.rs

```diff
@@ -484,23 +504,73 @@ where
     Ok(())
 }
 
+fn diff_with_cached<S>(settings: &S, path: &Path) -> bool
+where
+    S: Serialize,
+{
+    fn diff_with_cached_inner<S>(cached_settings: &S, path: &Path) -> Result<bool, DiffError>
+    where
+        S: Serialize,
+    {
+        let mut file = OpenOptions::new().read(true).open(path)?;
+        let mut buffer = String::new();
+        file.read_to_string(&mut buffer)?;
+        let s = serde_json::to_string(cached_settings)?;
+        let s = Sha256::digest_str(&s);
+        let encoded = base64::encode(&s);
+        if encoded == buffer {
+            debug!("Config state matches supplied config.");
+            Ok(false)
+        } else {
+            Ok(true)
+        }
+    }
+
+    match diff_with_cached_inner(settings, path) {
+        Ok(result) => result,
+
+        Err(err) => {
+            log_failure(Level::Debug, &err);
+            debug!("Error reading config backup.");
+            true
+        }
+    }
+}
+
+#[derive(Debug, Fail)]
+#[fail(display = "Could not load settings")]
+pub struct DiffError(#[cause] Context<Box<dyn std::fmt::Display + Send + Sync>>);
+
+impl From<std::io::Error> for DiffError {
+    fn from(err: std::io::Error) -> Self {
+        DiffError(Context::new(Box::new(err)))
+    }
+}
+
+impl From<serde_json::Error> for DiffError {
+    fn from(err: serde_json::Error) -> Self {
+        DiffError(Context::new(Box::new(err)))
+    }
+}
+
 fn check_settings_state<M, C>(
     subdir_path: PathBuf,
     filename: &str,
-    settings: &Settings<DockerConfig>,
+    settings: &M::Settings,
     runtime: &M,
     crypto: &C,
     tokio_runtime: &mut tokio::runtime::Runtime,
 ) -> Result<(), Error>
 where
-    M: ModuleRuntime,
-    <M as ModuleRuntime>::RemoveAllFuture: 'static,
+    M: MakeModuleRuntime + ModuleRuntime,
```

Comment:
The `ModuleRuntime` bound is needed here because this function calls `reconfigure` which in turn calls `ModuleRuntime::remove_all`. I was however able to remove the `RemoveAllFuture` bound and move the `'static` bound here.

Reply:
No. The point is that MakeMR and MR are two different traits, and one produces the other. Just because all our impls happen to implement both on the same type doesn't mean these functions should *require* both to be implemented on the same type.

Everywhere that you use M as a MR you should be using `<M as MakeModuleRuntime>::ModuleRuntime`.

Reply:
Understood.